peps/pep-0734.rst

```diff
@@ -0,0 +1,800 @@
+   assert ns['spam'] is True, ns
+
+* It propagates any uncaught exception from the code it ran.
+  The exception is raised from the ``exec()`` call in the thread
+  that originally called ``exec()``.
+
+The only difference is that, rather than propagate the uncaught
+exception directly, ``Interpreter.exec()`` raises an
+``interpreters.RunFailedError`` with a snapshot
+(``traceback.TracebackException``) of the uncaught exception
+(including its traceback) as the ``__cause__``.  That means if the
+``RunFailedError`` isn't caught then the full traceback of the
+propagated exception, including details about syntax errors, etc.,
+will be displayed.  Having the full traceback is particularly useful
+when debugging.
+
+Communicating Between Interpreters
+----------------------------------
+
+The module introduces a basic communication mechanism through special
+queues.
+
+There are ``interpreters.Queue`` objects, but they only proxy
+the actual data structure: an unbounded FIFO queue that exists
+outside any one interpreter.  These queues have special accommodations
+for safely passing object data between interpreters, without violating
+interpreter isolation.  This includes thread-safety.
+
+As with other queues in Python, for each "put" the object is added to
+the back and each "get" pops the next one off the front.  Every added
+object will be popped off in the order it was pushed on.
+
+Only objects that are specifically supported for passing
+between interpreters may be sent through a ``Queue``.  Note that the
+actual objects aren't sent, but rather their underlying data.
+However, the popped object will still be identical to the original.
+See `Shareable Objects`_.
+
+The module defines the following functions:
+
+* ``create_queue(maxsize=0) -> Queue``
+      Create a new queue.  If the maxsize is zero or negative then the
+      queue is unbounded.
+
+Queue Objects
+-------------
+
+``interpreters.Queue`` objects act as proxies for the underlying
+cross-interpreter-safe queues exposed by the ``interpreters`` module.
+
+``Queue`` implements all the methods of ``queue.Queue`` except for
+``task_done()`` and ``join()``, hence it is similar to
+``asyncio.Queue`` and ``multiprocessing.Queue``.
+
+Attributes and methods:
+
+* ``id``
+      (read-only) The target queue's ``QueueID`` object.
+      It is an int-like object with some internal bookkeeping.
+
+* ``maxsize``
+      Number of items allowed in the queue.
+
+* ``__hash__()``
+      Return the hash of the queue's ``id``.  This is the same
+      as the hash of the ID's integer value.
+
+* ``__eq__(other)``
+      Return ``other is self``.
+
+* ``empty()``
+      Return ``True`` if the queue is empty, ``False`` otherwise.
+
+* ``full()``
+      Return ``True`` if there are ``maxsize`` items in the queue.
+
+      If the queue was initialized with ``maxsize=0`` (the default),
+      then ``full()`` never returns ``True``.
+
+* ``qsize()``
+      Return the number of items in the queue.
```

Comment:
Maybe mention that these three are snapshots and their value can change any moment? E.g. qsize() may return 1 and yet get() could block -- if someone else snatched the item out of the queue before you.